resolver/dns_resolver_cache.ml

```diff
@@ -278,9 +278,18 @@ let handle_query t ~rng ip_proto ts (qname, qtype) =
     `Reply (flags, data), t
   | `Query name, t ->
     Logs.info (fun m -> m "query for %a (%a): %a" Domain_name.pp qname Packet.Question.pp_qtype qtype Domain_name.pp name);
-    let zone, name', types, ip, t = resolve t ~rng ip_proto ts name qtype in
+    (* DS should be requested at the parent *)
+    let name', recover =
+      if Domain_name.count_labels name > 1 && qtype = `K (Rr_map.K Ds) then
+        let n' = Domain_name.drop_label_exn name in
```

Comment:
I'm not sure about this. The parent zone is not necessarily the domain without one label, e.g. if example.org delegates bar.foo.example.org then the parent of bar.foo.example.org is not foo.example.org. Maybe I am missing something; I will have to read the context.

Reply:
True, the parent is not necessarily the zone cut, but this does not matter since DS records are only at the zone cuts. Take `example.com` and `cut.example.com` as zone cuts, then:
- query for DS of example.com should be done at the com nameservers
- query for DS of cut.example.com should be done at the example.com nameservers
- query for DS of foo.example.com does not exist (and is now done at the example.com nameservers)
- query for DS of bar.cut.example.com does not exist (but is done at the cut.example.com nameservers)